Readme updates (#222)
kubernetes-sigs/windows-testing: README.md

# Testing Kubernetes for Windows Clusters

> This repo is a collection of scripts, containers, and documentation needed to run Kubernetes end to end tests on clusters with Windows worker nodes.

Reviewer: ```suggestion
This repo is a collection of scripts, containers, and documentation needed to help run Kubernetes end to end tests on clusters with Windows worker nodes.
```

- It is maintained by [sig-windows](https://github.com/kubernetes/community/tree/master/sig-windows).

> - It leverages the existing upstream e2e tests, which live in Kubernetes.

Reviewer 2: ```suggestion
- It leverages the existing upstream e2e tests, which live in [Kubernetes repository](https://github.com/kubernetes/kubernetes/tree/master/test/e2e).
```

The rest of this page has a summary of those steps tailored for testing clusters that have Windows nodes.

## Building the tests

Testing windows is not different from regular testing of k8s clusters.  If you have never used a tool such as *e2e.test* or *sonobuoy* to run the Conformance (or other ginkgo based) test suites, then you should familiarize your self with that, first.

Some resources introducing the e2e test framework:

- https://kubernetes.io/blog/2019/03/22/kubernetes-end-to-end-testing-for-everyone/
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md#building-kubernetes-and-running-the-tests

Now, assuming you are able to build an e2e.test binary, we'll proceed with how windows the tests work.

**See the Questions section at the bottom of this doc** for pointers on how to build the e2e tests. 

## Running the e2e.test binary on a windows enabled cluster

If you already have a cluster, you will likely just need to build e2e.test, and run it with windows options.

> ### Method 1: Using an existing cluster

Reviewer: I think we should mention that you need to taint/label all linux nodes with node-role.kubernetes.io/master=NoSchedule.

This gets done by running this script on the master node during cluster proviosing
https://github.com/kubernetes-sigs/windows-testing/blob/c51f92045741c311d3e9a07eb00de79efddbb18f/extensions/master_extension/v1/win-e2e-master-extension.sh#L62

Author: ok done

All of the tests are built into the `e2e.test` binary, which you can as a standalone binary to test an existing cluster. 
If you don't know how to build `e2e.test`, instructions for this are at the bottom of this doc.

Note that, when running these **You need to set the KUBE_TEST_REPO_LIST** environment variable when running the windows end to end tests, otherwise, windows images will not be pullable and all your tests will fail !

- There are a few important parameters that you need to use, including:

- `--provider=skeleton` - this will avoid using a cloud provider to provision new resources such as storage volumes or load balancers
- `--ginkgo.focus="..."` - this regex chooses what [Ginkgo](http://onsi.github.io/ginkgo/) tests to run.
- `--node-os-distro="windows"` - some test cases have different behavior on Linux & Windows. This tells them to test Windows.
- `--ginkgo.skip="..."` - this regex chooses what tests to skip
- If you're not sure what test cases will run, add `--gingkgo.dryRun=true` and it will give a list of test cases selected without actually running them.

> - `e2e.test` also needs a few environment variables set to connect to the cluster, and choose the right test container images. Here's an example:

Reviewer 2: looks like the example go moved.  This particular line might be better off in the next section

Author: fixed

### Running the full suite of tests:

You can thus run all of the `[SIG-Windows]` tests like so:

```
export KUBECONFIG=path/to/kubeconfig
curl https://raw.githubusercontent.com/kubernetes-sigs/windows-testing/master/images/image-repo-list -o repo_list
export KUBE_TEST_REPO_LIST=$(pwd)/repo_list
./e2e.test --provider=skeleton --ginkgo.noColor --ginkgo.focus="\[sig-windows\]" --node-os-distro="windows"
```

### Running the test-grid tests exactly as run in CI

> The full list of what is run for TestGrid is in the [sig-windows-config.yaml](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/sig-windows/sig-windows-config.yaml) after `--test-args`. You can copy the parameters there for a full test to pass.

Reviewer 2: ```suggestion
The full list of what is run for TestGrid is in the [sig-windows-config.yaml](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/sig-windows/sig-windows-config.yaml) after `--test_args`. You can copy the parameters there for a full test to pass.  There are different parameters according to the type of test job it is (i.e there are parallel and serial test runs which exercise different set of tests)
```

> You can run these similarly to the above tests, just modifying the ginkgo.focus flag like so, and adding the `--ginkgo-skip` flag so as to skip a few tests which are known to fail.  This is thus suitable for CI deploymeents.

Reviewer 2: ```suggestion
You can run these similarly to the above tests, just modifying the ginkgo.focus flag like so, and adding the `--ginkgo-skip` flag to skip tests which are known to fail.  This is thus suitable for CI deployments.
```

Could you clarify `This is thus suitable for CI deployments.`?

```

> --ginkgo.focus=\\[Conformance\\]|\\[NodeConformance\\]|\\[sig-windows\\]|\\[sig-apps\\].CronJob --ginkgo.skip=\\[LinuxOnly\\]|\\[k8s.io\\].Pods.*should.cap.back-off.at.MaxContainerBackOff.\\[Slow\\]\\[NodeConformance\\]|\\[k8s.io\\].Pods.*should.have.their.auto-restart.back-off.timer.reset.on.image.update.\\[Slow\\]\\[NodeConformance\\]"

Reviewer 2: Maybe add a statement that points to this being an example?

The rest of this page has a summary of those steps tailored for testing clusters that have Windows nodes.

## Building the tests

Testing windows is not different from regular testing of k8s clusters.  If you have never used a tool such as *e2e.test* or *sonobuoy* to run the Conformance (or other ginkgo based) test suites, then you should familiarize your self with that, first.

Some resources introducing the e2e test framework:

- https://kubernetes.io/blog/2019/03/22/kubernetes-end-to-end-testing-for-everyone/
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md#building-kubernetes-and-running-the-tests

Now, assuming you are able to build an e2e.test binary, we'll proceed with how windows the tests work.

**See the Questions section at the bottom of this doc** for pointers on how to build the e2e tests. 

## Running the e2e.test binary on a windows enabled cluster

If you already have a cluster, you will likely just need to build e2e.test, and run it with windows options.

### Method 1: Using an existing cluster

> - FIRST you need to taint/label all linux and master nodes with `node-role.kubernetes.io/master=NoSchedule`... this way, any `windows` specific tests are gauranteed to run on Windows nodes.

Reviewer 3: s/gauranteed/guaranteed

All of the tests are built into the `e2e.test` binary, which you can as a standalone binary to test an existing cluster. 
If you don't know how to build `e2e.test`, instructions for this are at the bottom of this doc.

> Note that, when running these **You need to set the KUBE_TEST_REPO_LIST** environment variable when running the windows end to end tests, otherwise, windows images will not be pullable and all your tests will fail !

Reviewer 3: Is this really needed?

Reviewer 2: Right now it is required since the windows test images do not live in gcr repository.  There is work in progress for this and should be lit up in 1.21. It will still be required for tests running against 1.20 and earlier as only the new 1.21 tests images will be promoted I believe.

> ## Building Test Images

Reviewer: This section about Building Test Images might be out of date.
@claudiubelu has been working very hard to move the building of the windows test images into kubernetes/kubernetes.
I don't think that transition is fully complete yet tho...

Author: filed this https://github.com/kubernetes-sigs/windows-testing/issues/225  so we can maybe iterate that in a next patch ?

[images/](images/README.md) - has all of the container images used in e2e test passes and the scripts to build them. They are replacement Windows containers for those in [kubernetes/test/images](https://github.com/kubernetes/kubernetes/tree/master/test/images)

# Questions

## Is there an equivalent to a Conformance test suite for windows ? 

> There is not yet a formal equivalent.  The purpose of this repo is to store existing tests which sig-windows currently runs, and to provide CI signal for windows to upstream Kubernetes.  If you run the e2e.test suite with the `sig-windows` ginkgo filter and the `--ginkgo.dryRun` option, you'll see the list of tests which are currently in upstream.  these are listed below.  In order to implement Conformance tests for windows - we would need to, on a large scale, implement windows specific logic for certain things.  There has been made some progress on this (for example, with the agnhost container), but by in large the ability to run most tests on windows/linux interchangeably doesn't yet exist.

Reviewer 2: @adelina-t do you have any thing you could link to for the plans/progress for conformance tests here?

## Can I run sonobuoy as a way to test windows Conformance ? 

Sonobuoy is currently not aware of windows/linux taints, and thus might not currently run the end to end test suites on a mixed windows/linux cluster, and also does not yet have an option for enabling specific node-os behaviour (via the `node-os-distro` flag).

## How many tests run ? 

> As of Kubernetes 1.19, there are currently ~15 tests in upstream Kubernetes (1.19) which specifically target the windows feature.

Reviewer 2: I think @adelina-t had a count of 28 when we chatted last week?

Author: right now on master its 15, i think ? 

```
➜  kubernetes git:(netpol-impl2-follow-on-empty-policies) ✗  KUBECONFIG=/Users/jayunit100/.kube/config .//_output/local/go/bin/e2e.test --ginkgo.focus="Windows" --ginkgo.dryRun=true --dump-logs-on-failure=false  
I1203 07:23:45.222439   63229 test_context.go:457] Tolerating taints "node-role.kubernetes.io/master" when considering if nodes are ready
Dec  3 07:23:45.222: INFO: The --provider flag is not set. Continuing as if --provider=skeleton had been used.
I1203 07:23:45.222886   63229 e2e.go:129] Starting e2e run "dc1b8d5e-5cac-4801-b9b1-7e148d78bb97" on Ginkgo node 1
{"msg":"Test Suite starting","total":15,"completed":0,"skipped":0,"failed":0}
Running Suite: Kubernetes e2e suite
===================================
Random Seed: 1606998224 - Will randomize all specs
Will run 15 of 5337 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] Kubelet-Stats [Serial] Kubelet stats collection for Windows nodes when running 10 pods 
  should return within 10 seconds
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/kubelet_stats.go:44
•{"msg":"PASSED [sig-windows] [Feature:Windows] Kubelet-Stats [Serial] Kubelet stats collection for Windows nodes when running 10 pods should return within 10 seconds","total":15,"completed":1,"skipped":704,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] SecurityContext 
  should ignore Linux Specific SecurityContext if set
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/security_context.go:74
•{"msg":"PASSED [sig-windows] [Feature:Windows] SecurityContext should ignore Linux Specific SecurityContext if set","total":15,"completed":2,"skipped":820,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] Density [Serial] [Slow] create a batch of pods 
  latency/resource should be within limit when create 10 pods with 0s interval
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/density.go:66
•{"msg":"PASSED [sig-windows] [Feature:Windows] Density [Serial] [Slow] create a batch of pods latency/resource should be within limit when create 10 pods with 0s interval","total":15,"completed":3,"skipped":908,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] Windows volume mounts  check volume mount permissions 
  container should have readOnly permissions on emptyDir
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/volumes.go:65
•{"msg":"PASSED [sig-windows] Windows volume mounts  check volume mount permissions container should have readOnly permissions on emptyDir","total":15,"completed":4,"skipped":1338,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] Memory Limits [Serial] [Slow] Allocatable node memory 
  should be equal to a calculated allocatable memory value
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/memory_limits.go:49
•{"msg":"PASSED [sig-windows] [Feature:Windows] Memory Limits [Serial] [Slow] Allocatable node memory should be equal to a calculated allocatable memory value","total":15,"completed":5,"skipped":1404,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] GMSA Kubelet [Slow] kubelet GMSA support when creating a pod with correct GMSA credential specs 
  passes the credential specs down to the Pod's containers
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/gmsa_kubelet.go:43
•{"msg":"PASSED [sig-windows] [Feature:Windows] GMSA Kubelet [Slow] kubelet GMSA support when creating a pod with correct GMSA credential specs passes the credential specs down to the Pod's containers","total":15,"completed":6,"skipped":2207,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] SecurityContext 
  should not be able to create pods with unknown usernames
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/security_context.go:46
•{"msg":"PASSED [sig-windows] [Feature:Windows] SecurityContext should not be able to create pods with unknown usernames","total":15,"completed":7,"skipped":2317,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] GMSA Full [Serial] [Slow] GMSA support 
  works end to end
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/gmsa_full.go:86
•{"msg":"PASSED [sig-windows] [Feature:Windows] GMSA Full [Serial] [Slow] GMSA support works end to end","total":15,"completed":8,"skipped":2828,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] SecurityContext 
  should override SecurityContext username if set
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/security_context.go:60
•{"msg":"PASSED [sig-windows] [Feature:Windows] SecurityContext should override SecurityContext username if set","total":15,"completed":9,"skipped":2916,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] Hybrid cluster network for all supported CNIs 
  should have stable networking for Linux and Windows pods
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/hybrid_network.go:52
•{"msg":"PASSED [sig-windows] Hybrid cluster network for all supported CNIs should have stable networking for Linux and Windows pods","total":15,"completed":10,"skipped":3170,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] Services 
  should be able to create a functioning NodePort service for Windows
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/service.go:42
•{"msg":"PASSED [sig-windows] Services should be able to create a functioning NodePort service for Windows","total":15,"completed":11,"skipped":3364,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] SecurityContext 
  should be able create pods and run containers with a given username
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/security_context.go:37
•{"msg":"PASSED [sig-windows] [Feature:Windows] SecurityContext should be able create pods and run containers with a given username","total":15,"completed":12,"skipped":4364,"failed":0}
SSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] Cpu Resources [Serial] Container limits 
  should not be exceeded after waiting 2 minutes
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/cpu_limits.go:41
•{"msg":"PASSED [sig-windows] [Feature:Windows] Cpu Resources [Serial] Container limits should not be exceeded after waiting 2 minutes","total":15,"completed":13,"skipped":4378,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] Windows volume mounts  check volume mount permissions 
  container should have readOnly permissions on hostMapPath
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/volumes.go:74
•{"msg":"PASSED [sig-windows] Windows volume mounts  check volume mount permissions container should have readOnly permissions on hostMapPath","total":15,"completed":14,"skipped":4817,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-windows] [Feature:Windows] Memory Limits [Serial] [Slow] attempt to deploy past allocatable memory limits 
  should fail deployments of pods once there isn't enough memory
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/windows/memory_limits.go:55
•{"msg":"PASSED [sig-windows] [Feature:Windows] Memory Limits [Serial] [Slow] attempt to deploy past allocatable memory limits should fail deployments of pods once there isn't enough memory","total":15,"completed":15,"skipped":5247,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS{"msg":"Test Suite completed","total":15,"completed":15,"skipped":5322,"failed":0}

Ran 15 of 5337 Specs in 0.144 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 5322 Skipped
PASS
```
